rocky/reports/report_types/mail_report/report.py

```diff
@@ -51,21 +51,21 @@ def generate_data(self, input_ooi: str, valid_time: datetime) -> dict[str, Any]:
 
             number_of_spf -= (
                 1
-                if any(
+                if list(
```

Comment:
What have you done or searched to fix this Mypy error? A quick lookup reveals that in this context it indeed requires a type guarded/ type narrowing function. An even simpler approach would be to rewrite the filter with lambda to a generator expression (`any(( expr for item in finding_types[...]))`)

Reply:
I don't think these kinds of comments really put a good tone on PR reviews. With these words you are:
1. Suggesting I didn't try anything to resolve it (by asking what I tried)
2. Saying that it was actually "quick" and easy to look the issue up and resolve it.

But if you must know I looked at:
1. Narrowing the return type of the `group_finding_types_by_source()` method, of which I expected the error to move towards perhaps a mismatch with the return type of the `group_by_source()` method, which it did not
2. Type hinting the variables, without any change
3. How it could be possible in general that the type error doesn't occur when we call `list()` but does occur when we call `any()`

At which point I figured this is perhaps not worth putting too much time into since I think this logic could be refactored in other ways and this PR focuses around other major API changes. 

P.S.
Isn't it strange that mypy doesn't complain about the id field now that it's wrapped in the plain `any()`?

Reply:
Yes, I also don't really understand it from a quick look, maybe it is because mypy is still lacking in some areas. We also don't have all strict options turned on and the models are still excluded from mypy here (my octopoes typing PR will fix that) which might also cause this, not sure about that.

While in general we should use `any` and generators because they are better performance wise, I think that

1. The current code already does it in the suboptimal way, so I don't see this as a blocker for merging this PR.
2. I think the way we use filters, generators, count findings, etc. in the reports are symptoms that we aren't using the right datastructures and we should first focus on those things. For example use a single query to fetch all relevant findings of hostnames for multiple reports in one go and have it return a set of findingtype ids we can use `not in` on without needing a generator or filter/lambda at all. Or if we only need to know the total number of each findingtype we might be able to just let the database count them.

Reply:
> 3. How it could be possible in general that the type error doesn't occur when we call `list()` but does occur when we call `any()`

It's a bit strange indeed, it seems as if Mypy is picking the wrong overload of filter's initializer. It has been reported a while ago but hasn't been fixed. So far I've only seen people bypassing this by using a generator expression in the `any` instead.

> At which point I figured this is perhaps not worth putting too much time into since I think this logic could be refactored in other ways and this PR focuses around other major API changes.

Agree

> 2\. I think the way we use filters, generators, count findings, etc. in the reports are symptoms that we aren't using the right datastructures and we should first focus on those things. For example use a single query to fetch all relevant findings of hostnames for multiple reports in one go and have it return a set of findingtype ids we can use `not in` on without needing a generator or filter/lambda at all. Or if we only need to know the total number of each findingtype we might be able to just let the database count them.

Yep, that's some food for thoughts